lib/MonadicRewrite.thy

```diff
@@ -14,24 +15,100 @@ imports
   LemmaBucket
 begin
 
-definition
-  monadic_rewrite :: "bool \<Rightarrow> bool \<Rightarrow> ('a \<Rightarrow> bool)
-                       \<Rightarrow> ('a, 'b) nondet_monad \<Rightarrow> ('a, 'b) nondet_monad \<Rightarrow> bool"
-where
- "monadic_rewrite F E P f g \<equiv> \<forall>s. P s \<and> (F \<longrightarrow> \<not> snd (f s))
-           \<longrightarrow> (E \<longrightarrow> f s = g s)
-            \<and> (\<not> E \<longrightarrow> fst (g s) \<subseteq> fst (f s) \<and> (snd (g s) \<longrightarrow> snd (f s)))"
+definition monadic_rewrite ::
+  "bool \<Rightarrow> bool \<Rightarrow> ('s \<Rightarrow> bool) \<Rightarrow> ('s, 'a) nondet_monad \<Rightarrow> ('s, 'a) nondet_monad \<Rightarrow> bool"
+  where
+   "monadic_rewrite F E P f g \<equiv>
+      \<forall>s. P s \<and> (F \<longrightarrow> \<not> snd (f s))
+          \<longrightarrow> (E \<longrightarrow> f s = g s) \<and> (\<not> E \<longrightarrow> fst (g s) \<subseteq> fst (f s) \<and> (snd (g s) \<longrightarrow> snd (f s)))"
 
+lemma monadic_rewrite_inst:
+  "monadic_rewrite F E P f g \<Longrightarrow> monadic_rewrite F E P f g"
+  by assumption
+
+lemma monadic_rewrite_refl:
+  "monadic_rewrite F E \<top> f f"
+  by (simp add: monadic_rewrite_def)
+
+lemma monadic_rewrite_impossible:
+  "monadic_rewrite F E \<bottom> f g"
+  by (clarsimp simp: monadic_rewrite_def)
+
+lemma monadic_rewrite_trans:
+  "\<lbrakk> monadic_rewrite F E P f g; monadic_rewrite F E Q g h \<rbrakk>
+   \<Longrightarrow> monadic_rewrite F E (P and Q) f h"
+  by (auto simp add: monadic_rewrite_def)
+
+lemma monadic_rewrite_trans_dup:
+  "\<lbrakk> monadic_rewrite F E P f g; monadic_rewrite F E P g h \<rbrakk>
+   \<Longrightarrow> monadic_rewrite F E P f h"
+  by (auto simp add: monadic_rewrite_def)
```

Comment:
Minor nitpick: it'd be neat to do the proof of this one in terms of `monadic_rewrite_trans` so that it's clear that the first is fundamental and this one here is derivable.

Reply:
Done as requested.